monitor/monitor.go

```diff
@@ -0,0 +1,260 @@
+	if err := m.start(monitorName, monitorConfig); err != nil {
+		log.Errorf("%v: Unable to start new monitor '%v': %v", m.Identifier, monitorName, err.Error())
+		return fmt.Errorf("Unable to start new monitor %v: %v", monitorName, err.Error())
+	}
+
+	log.Debugf("%v: Successfully started new monitor %v!", m.Identifier, monitorName)
+
+	return nil
+}
+
+// Perform the actual stop of a given monitor; update running monitor slice
+func (m *Monitor) stop(monitorName string) error {
+	// Stop the given monitor
+	m.runningMonitorLock.Lock()
+	defer m.runningMonitorLock.Unlock()
+
+	// Double check
+	if _, ok := m.runningMonitors[monitorName]; !ok {
+		return fmt.Errorf("%v: Unable to stop monitor '%v' - this monitor is NOT running!", m.Identifier, monitorName)
+	}
+
+	// Stop the actual check
+	m.runningMonitors[monitorName].Stop()
+
+	// Remove it from runningMonitors
+	delete(m.runningMonitors, monitorName)
+
+	return nil
+}
+
+// Perform the actual start of a monitor; update running monitor slice
+func (m *Monitor) start(monitorName string, monitorConfig *MonitorConfig) error {
+	// Let's be overly safe and ensure this monitor type exists
+	if _, ok := m.SupportedMonitors[monitorConfig.Type]; !ok {
+		return fmt.Errorf("%v: No such monitor type found '%v'", m.Identifier, monitorConfig.Type)
+	}
+
+	// Create a new monitor instance
+	newMonitor := m.SupportedMonitors[monitorConfig.Type](
+		&RootMonitorConfig{
+			Name:           monitorName,
+			GID:            util.RandomString(GOROUTINE_ID_LENGTH, false),
+			Config:         monitorConfig,
+			MessageChannel: m.MessageChannel,
+			StopChannel:    make(chan bool, 1),
+			Ticker:         time.NewTicker(time.Duration(monitorConfig.Interval)),
+		},
+	)
+
+	m.runningMonitorLock.Lock()
+	defer m.runningMonitorLock.Unlock()
+
+	// Add monitor to runningMonitors
+	m.runningMonitors[monitorName] = newMonitor
+
+	// Launch the monitor
+	go m.runningMonitors[monitorName].Run()
+
+	return nil
+}
+
+// Determine if given `monitorName` is in `runningMonitors`
+func (m *Monitor) monitorRunning(monitorName string) bool {
+	m.runningMonitorLock.Lock()
+	defer m.runningMonitorLock.Unlock()
+
+	for k, _ := range m.runningMonitors {
+		if k == monitorName {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Ensure that the monitoring config is valid
+func (m *Monitor) validateMonitorConfig(monitorConfig *MonitorConfig) error {
+	// TODO: HTTP* validation
+
+	if monitorConfig.Interval.String() == "0s" {
+		return errors.New("'Interval' must be > 0s")
+	}
+
+	if monitorConfig.CriticalThreshold == 0 {
+		return errors.New("'CriticalThreshold' must be non-zero")
+	}
+
+	// TODO: Logic for this should be changed/fixed at some point
+	if monitorConfig.WarningThreshold > monitorConfig.CriticalThreshold {
+		return errors.New("'WarningThreshold' cannot be larger than 'CriticalThreshold'")
+	}
+
+	// TODO: It should be possible to NOT have a WarningAlerter setting (and just
```

Comment:
This seems important. Seems like it should be possible to NOT specify a warning threshold and just have a critical threshold.